### ISSUE TITLE
MSUnmerged: Reverse mapping of data present at /store/unmerged to a wrokflow name

### ISSUE THREAD

**Issue by todor-ivanov** [dmwm/WMCore#10969]
**Impact of the new feature**
MicroServices: MSUnmerged

**Is your feature request related to a problem? Please describe.**
While looking into different sites which are suffering too much data locked at the site in their `/store/unmerged/` branch of the tree we and P&R started to ask the question which is the workflow locking the data there, for how long it was at  the site and if there is a  reason for it to stay there etc. This  leads to a need for the reverse mapping:

`/store/unmerged/Directory` -> `Workflow description at Request Magnager`  

**Describe the solution you'd like**
Two Possible solutions do exist as discussed shortly in the comments here [1]:

* The set of information parameters needed to me constantly present and preserved at the MSUnmerged database for every data recorded in the site's `/store/unmerged/` branch
* A REST API to the `MSUnmerged` service, which would do that reverse mapping on a per call basis and just returning a workflow name. Something like:

https://cmsweb-testbed.cern.ch/ms-unmerged/data/info?rse=T2_US_Florida&lfn=/store/unmerged/RunIISummer20UL16NanoAODv9/QCD_HT300to500_TuneCP5_PSWeights_13TeV-madgraphMLM-pythia8/NANOAODSIM/106X_mcRun2_asymptotic_v17-v1&detail=True

 From those two later one sounds more realistic, because the former  would involve calls to several other services like DBS and RequestManager and would increase the records in the database in orders of magnitude. 

[1]
https://github.com/dmwm/WMCore/issues/10686

**Describe alternatives you've considered**
Not discussed

**Additional context**
N/A

**Comment by amaltaro** [comment]
This has been marked as one of the top priority items for this quarter, by the P&R team. So moving it to the Q1, but it's still behind the final bug-fixes and functional changes required in MSUnmerged

**Comment by todor-ivanov** [comment]
Thanks  @amaltaro

**Comment by vkuznet** [comment]
@todor-ivanov I'm interested to learn more about this issue, since it seems like you're trying to do DAS like query, i.e. knowing one meta-data find another one associated with it. In order to have clear picture do you mind to post the data relations among different services and how you envision to traverse the data lookup, e.g.
- user provide `/store/unmerged/Directory` parameter (and/or site)
- we need to query service A to get information B
- using B we need to query service X to get C
- finally using C we can query service Y to get workflow name
In other words, I want to know which services store what and which steps are required (in plain English) to make this reverse look-up. Once we'll outline which services needs to be involved and which information is required to do the traverse look-up it would be much easier to understand how to implement such actions.

**Comment by todor-ivanov** [comment]
Hi @vkuznet  initially when discussing this functionality with @amaltaro we were planning to do the following sequence of actions:
* Parse the LfnBase from the REST call
* Make a query to DBS in order to find which dataset the files from this Lfn belong to
* Make a query to Request Manager to match the workflow in question by setting a filter for all workflows having the so found dataset  as output dataset   

But now as I dig deeper on how actually the `OutputModulesLFNBases` are created we can even skip the call to DBS. What I find is a solid pattern in this (@amaltaro  please correct me if I am wrong here), which pattern is as follows (e.g. taking this workflow, currently in the system as an example [pdmvserv_task_HIG-RunIISummer20UL16MiniAOD-01322__v1_T_220504_141651_2512](https://cmsweb.cern.ch/reqmgr2/fetch?rid=pdmvserv_task_HIG-RunIISummer20UL16MiniAOD-01322__v1_T_220504_141651_2512)):
* Unmerged area lfn base for all output modules:
```
"OutputModulesLFNBases": [
    "/store/unmerged/RunIISummer20UL16MiniAOD/VBF_LFV_HToEMu_M140_TuneCH3_13TeV_powheg_herwig7/MINIAODSIM/106X_mcRun2_asymptotic_v13-v1"
  ]
```
Which is created by the following model: `/<UnmergedLFNBase>/<Campaign>/<PrimaryDataset>/<DataTier>/<ProcessingString>-<version>/`
   * `"UnmergedLFNBase": "/store/unmerged/"`
   * `"Campaign": "RunIISummer20UL16MiniAOD"  | "AcquisitionEra": "RunIISummer20UL16MiniAOD"`
   * `"ProcessingString": "106X_mcRun2_asymptotic_v13" | "GlobalTag": "106X_mcRun2_asymptotic_v13"`
   * `"PrimaryDataset": "VBF_LFV_HToEMu_M140_TuneCH3_13TeV_powheg_herwig7"`

* While the output dataset is:
```
  "OutputDatasets": [
    "/VBF_LFV_HToEMu_M140_TuneCH3_13TeV_powheg_herwig7/RunIISummer20UL16MiniAOD-106X_mcRun2_asymptotic_v13-v1/MINIAODSIM"
  ]
```
And seems to be created by the following model: `/<PrimaryDataset>/<Campaign>-<ProcessingString>-<version>/<DataTier>/`

So I think we are having all the information needed from the output module lsfnbases  so taht we can create the proper filters to request manager such that  we can uniquely identify the workflow producing the output in question.

**Comment by todor-ivanov** [comment]
Hi @vkuznet , 
Just tot add something I was thinking about mentioning yesterday, but in the course of writing my long comment I forgot an important detail. 
The approach with quering DBS for finding output datasets starting by parsing a record in `/store/unmerged/` at a site (which record must have its one to one equivalent with a record in a workflow's `OutputModulesLFNBases`) won't work for everything we have at the site. Because we have steps/tasks with `KeepOutput=False`. Which means no metadata from them would ever reach DBS. Shortly speaking: 

 * `set(All sites' /store/unmerged/.. records) == set(All workflows'  OutputModulesLFNBases deinitions)`
 * `set(All output datasets from all workflows) == set(All records in DBS)`
 But:
* `set(All workflows' OutputModulesLFNBases defenitions) > set(All output datasets from all workflows)` 

@amaltaro  please do add your input here. It would be good to correct me or add a missing detail as the one above on an early stage.

**Comment by amaltaro** [comment]
I think the proper way to map unmerged files to a workflow depends solely on the workflow high level description (json data). E.g.:
https://cmsweb.cern.ch/reqmgr2/fetch?rid=amaltaro_TC_HC_HG2202_Val_220218_174025_9506

has
```
  "OutputModulesLFNBases": [
    "/store/unmerged/CMSSW_12_2_0/TTbar_14TeV/GEN-SIM-RECO/Reco_2021_TC_HC_HG2202_Val_Alanv4-v11",
    "/store/unmerged/CMSSW_12_2_0/TTbar_14TeV/GEN-SIM-RAW/Digi_2021_TC_HC_HG2202_Val_Alanv4-v11",
    "/store/unmerged/CMSSW_12_2_0/TTbar_14TeV/AODSIM/Reco_2021_TC_HC_HG2202_Val_Alanv4-v11",
    "/store/unmerged/CMSSW_12_2_0/TTbar_14TeV/GEN-SIM/TTbar_14TeV_GenSim_TC_HC_HG2202_Val_Alanv4-v11"
  ],
```

so, once we have an unmerged LFN, we need to strip the filename (and possible the hash code made out of the agent name (or agent number) and match that string against the `OutputModulesLFNBases` list.

It does require to parse every single workflow from the database though (~750k right now). So we might have to do something else to speed up this mapping, e.g.:
* fetch this information only for active workflows (which we could then use the wmstatsserver data cache
* create a CouchDB view yielding output LFN and workflow id

Identifying the pattern of use of such feature would be interesting as well. Is it supposed to be done by people checking files here and there? Is it supposed to be used by a service that will provide such information in a continuous mode? This is important to evaluate the impact on the system and make sure that whatever is made available scales in the long run.

**Comment by todor-ivanov** [comment]
Thanks @amaltaro I was also thinking about just following the pattern of mapping `OutputModulesLFNs` to workflow names  internally. If we could actually create the proper filters in `WmstatsServer` so that  we can query directly for workflows and get  only the `OutputmodulesLFNs` and a workflow name this could reduce the data returned alot. I think I can do that.

> create a CouchDB view yielding output LFN and workflow id

This could indeed speedup things. 

> Identifying the pattern of use of such feature would be interesting as well. 
The pattern[s] of use would be actually two: 
 * P&R wanted  to have the information constantly aggregated and a proper summary provided for stuck workflows.
 * We suggested to initially provide the interface  for debugging purposes  only - People querying the database for long living LFNs at a site so that they could identify if those belong to a stuck workflow or not.

**Comment by vkuznet** [comment]
thanks for sharing details since now it becomes more clear how to do it better. Here is some improvements I suggest:
- during creation of `OutputmodulesLFNs` whoever who creates this info can either perform stripping of LFN base, i.e. /store/unmerged, or creates its md5 hash and add to the document as `StoragePath`.
- then you may create index on this attribute in your underlying DB
- doing this way you should not force people to know actual LFNs since in fact the original request does not require it, i.e. you asked for `/store/unmerged/Directory -> Workflow description at Request Manager`. Therefore, you don't need to force people to know LFNs and can simplify API to accept only site and this `StoragePath`
- finally, you should adjust now description that only single service can do this job and then provide in ReqMgr2 new API which will take path and site as input parameters.
